### ISSUE TITLE
#12 yolov7 + jetson

### ISSUE THREAD

**Issue by mihara-shoko** [aifield/CV_News#12]
# yolov7実装紹介
+ github：yolov7 https://github.com/WongKinYiu/yolov7
+ ファーストリリース：2022/07

# 目的

+ オリジナルデータセットを使い、学習できるようにする。
+ yolov7をJetsonのDeepStream上で動かせるようにする。
+ yolov5n, yolov5s, yolov7-tinyの速さを比較する。

# 一言説明

+ yolov7は2022年7月に発表された物体検出モデル
+ Real-Time Object Detection on COCOでSoTA達成

<img width="400" alt="スクリーンショット 2022-12-11 14 40 35" src="https://user-images.githubusercontent.com/65023550/206888335-05aca56c-ed76-4857-9809-6dbd9f6a9384.png">

# JetsonとDeepStreamについて
## Jetson
+ NVIDIAが開発した小型コンピュータ。
+ NVIDIAのGPUが搭載されている。

<img width="300" alt="PXL_20221211_050420330" src="https://user-images.githubusercontent.com/65023550/206888562-6710308d-6a9e-4a88-9005-a7cff4ed1fd3.jpg">

## DeepStream
+ GStreamerをベースにしたストリーム分析ツールキット
+ AIを使ったビデオ解析、音声分析、画像分析のためのパイプラインを構築できる
+ C/C++またはPythonで開発できる（Graph ComposerというUIを使った開発もできるらしい）
+ 単純なパイプラインを作るためには便利だが、複雑なことをやろうとすると不便

<img width="600" alt="スクリーンショット 2022-12-11 15 03 53" src="https://user-images.githubusercontent.com/65023550/206888961-ac57008a-4b36-4ff4-a712-fb3204070770.png">

# 使ってみよう
yolov7-tiny, yolov5s, yolov5nを比較した。
## yolov5, yolov7の学習
yolov5とyolov7はほぼ同じように学習できる。  
train.ipynbを参照。  

### 学習
以下の条件でyolov5n, yolov5s, yolov7-tinyを学習した。  
yolov7は最新のリポジトリ 、yolov5はv6.2のリポジトリ を使用して学習した。
|  class  |  train data  |  valid data  |  input size  |  epoch  |  batch size  |  device  |
| ---- | ---- | ---- | ---- | ---- | ---- | ---- |
|  1<br>(person) |  1000枚<br>Open image dataset  |  333枚<br>Open image dataset  |  416  |  100  |  128  |  NVIDIA Tesla T4  |

学習にかかった時間  

|  model  |  所要時間  |  パラメータ数  |
| ---- | ---- | ---- |
|  yolov5n  |  1 h 46 min  |  1.9 M  |
|  yolov5s  |  1 h 51 min  |  7.2 M  |
|  yolov7-tiny  |  2 h 59 min  |  6.2 M  |

学習ログ  
+ mAP_0.5, mAP0.5:0.95, precision, recall

<img width="700" alt="スクリーンショット 2022-12-11 15 33 35" src="https://user-images.githubusercontent.com/65023550/206889875-7e56f641-8110-4bd3-8ddb-0b1a5d218175.png">

+ train loss

<img width="700" alt="スクリーンショット 2022-12-11 15 34 58" src="https://user-images.githubusercontent.com/65023550/206889902-d81df005-af02-4344-82ff-b914f8e02cb7.png">

+ val loss

<img width="700" alt="スクリーンショット 2022-12-11 15 35 55" src="https://user-images.githubusercontent.com/65023550/206889939-53ae1e56-68b3-4b4e-b0e3-08423cd2081e.png">

### 評価
以下の条件でyolov5n, yolov5s, yolov7-tinyの精度を評価した。  

|  test data  |  input size  |  conf thres  |  iou thres  |
| ---- | ---- | ---- | ---- |
|  500枚<br>Open image dataset |  416  |  0.25  |  0.45  |

評価結果  

|  model  |  Precision  |  Recall  |  mAP@0.5  |  パラメータ数  |
| ---- | ---- | ---- | ---- | ---- |
|  yolov5n  |  0.706  |  0.659  |  0.677  |  1.9 M  |
|  yolov5s  |  0.710  |  0.688  |  0.702  |  7.2 M  |
|  yolov7-tiny  |  0.713  |  0.725  |  0.650  |  6.2 M  |

<br>

## Jetson組込
サンプルアプリdeepstream-appを使ってカメラ映像 or 動画についてyolov7で推論を行えるようにする。

環境  
+ Jetson nano 4GB
+ Jetpack 4.5
+ Deepstream SDK 5.1

組込手順  
1. モデルのTensorRT化: モデルをDeepStreamで使える形式に変換する。デバイスに合わせてモデルを高速化する。
2. コンフィグファイルの設定: 前処理の設定などを行う。
3. パーサー作成: 後処理(NMS処理など)を行うファイルを作成する。
4. サンプルアプリ実行

### 1. モデルのTensorRT化
モデルをDeepStreamで使える形式に変換する。デバイスに合わせてモデルを高速化する。　　
yolov5と同様に、.pt -> .wts -> .engineの順で変換する。
.wtsファイルまではgoogle colabで変換し、.engineファイルへの変換はJetsonで行う。  
TensorRT.ipynb参照。

### 2. コンフィグファイルの設定
前処理の設定などを行う。　　

以下のリポジトリをダウンロードし、parserフォルダをJetsonの/opt/nvidia/deepstream/deepstream-6.0/sources/下に保存する。  
https://github.com/mihara-shoko/yolo_parser

必要があればconfig_infer_primary.txtを修正する。  
```
[property]
gpu-id=0
net-scale-factor=0.0039215697906911373
model-color-format=0 
model-engine-file=yolov7_tiny.engine # TensorRT化したモデルのパス
labelfile-path=labels.txt # ラベルファイルのパス
num-detected-classes=1 # クラス数
interval=0
gie-unique-id=1
process-mode=1
network-type=0
cluster-mode=4
maintain-aspect-ratio=1
parse-bbox-func-name=NvDsInferParseCustomYoloV5
custom-lib-path=nvdsinfer_custom_impl_Yolo/libnvdsinfer_custom_impl_Yolo.so

[class-attrs-all]
pre-cluster-threshold=0.25 # conf thres
```
前処理に関係する項目
|  項目  |  説明  |  yolov7の場合の設定値  |
| ---- | ---- | ---- |
|  net-scale-factor |  ピクセル値にかける値<br>モデルにはyのピクセル値が入力される。<br>y = net scale factor*(x-offset)  |  0.0039215697906911373<br>0-255を0-1のrangeにしている  |
|  offset |  ピクセル値から引く値<br>RGBそれぞれ別に設定できる。<br>モデルにはyのピクセル値が入力される。<br>y = net scale factor*(x-offset)  |  0:0:0<br>設定しなくて良い  |
|  model-color-format |  0:RGB<br>1:BGR<br>2:GRAY  |  0  |
|  maintain-aspect-ratio |  リサイズするときにアスペクト比を維持するかどうか  |  1  |

詳しくはNVIDIAのページを参照  
https://docs.nvidia.com/metropolis/deepstream/dev-guide/text/DS_plugin_gst-nvinfer.html

### 3. パーサー作成
後処理(NMS処理など)を行うファイルを作成する。

Jetsonの/opt/nvidia/deepstream/deepstream-6.0/sources/parserに移動する。  
nvdsinfer_custom_impl_Yolo/yololayer.hのクラス数(19行目)、input size(20, 21行目)を適宜変更する。

```nvdsinfer_custom_impl_Yolo/yololayer.h
#ifndef _YOLO_LAYER_H
#define _YOLO_LAYER_H

#include <vector>
#include <string>
#include "NvInfer.h"

namespace Yolo
{
    static constexpr int CHECK_COUNT = 3;
    static constexpr float IGNORE_THRESH = 0.1f;
    struct YoloKernel
    {
        int width;
        int height;
        float anchors[CHECK_COUNT * 2];
    };
    static constexpr int MAX_OUTPUT_BBOX_COUNT = 1000;
    static constexpr int CLASS_NUM = 1;
    static constexpr int INPUT_H = 416;  // yolov5's input height and width must be divisible by 32.
    static constexpr int INPUT_W = 416;

```

以下のコマンドでコンパイルする。

```
CUDA_VER=10.2 make -C nvdsinfer_custom_imple_Yolo
```
nvdsinfer_custom_impl_Yolo下生成した.soファイルがパーサーとなる。

### 4. サンプルアプリ実行
サンプルアプリを使ってyolov7で推論を行う。推論速度をyolov5n, yolov5s, yolov7と比較する。  

Jetsonの/opt/nvidia/deepstream/deepstream-6.0/sources/parserに移動する。
deepstream_app_config.txtのuriに読み込ませたい動画のパスを設定する。  
```
[source0]
enable=1
type=3
uri=file://../../samples/streams/sample_1080p_h264.mp4
num-sources=1
gpu-id=0
cudadec-memtype=0
```
以下のコマンドを実行し、アプリを実行する。

```
deepstream-app -c deepstream_app_config.txt
```

アプリ実行条件
|  input size  |  データ型  |  conf thres  |  iou thres  |
| ---- | ---- | ---- | ---- |
|  416 |  FP16  |  0.25  |  0.45  |

FPS

|  model  |  FPS  |  .ptモデルのパラメータ数  |
| ---- | ---- | ---- |
|  yolov5n  |  55  |  1.9 M  |
|  yolov5s  |  28  |  7.2 M  |
|  yolov7-tiny  |  37  |  6.2 M  |

# 利用した所感

+ yolov5を使ったことがある人なら、yolov7の学習と組み込みが同様の手順で行える。
+ yolov5n, yolov5s, yolov7-tinyの精度はあまり変わらなかった。
+ yolov5よりも推論速度が速いため、アプリ開発で使えそう。

# 参考資料

+ https://developer.nvidia.com/ja-jp/deepstream-sdk#:~:text=NVIDIA%20%E3%81%AE%20DeepStream%20SDK%20%E3%81%AF,OEM%20%E5%90%91%E3%81%91%E3%81%AE%E8%A3%BD%E5%93%81%E3%81%A7%E3%81%99%E3%80%82
+ https://docs.nvidia.com/metropolis/deepstream/dev-guide/text/DS_plugin_gst-nvinfer.html
